viam-cartographer.go

```diff
@@ -0,0 +1,988 @@
+			}
+
+			brownConrady, ok := props.DistortionParams.(*transform.BrownConrady)
+			if !ok {
+				return "", nil, errors.New("error getting distortion_parameters for slam service, " +
+					"only BrownConrady distortion parameters are supported")
+			}
+			if err := brownConrady.CheckValid(); err != nil {
+				return "", nil, errors.Wrapf(err, "error validating distortion_parameters for slam service")
+			}
+		}
+
+		cams = append(cams, cam)
+
+		// If there is a second camera, it is expected to be depth.
+		if len(svcConfig.Sensors) > 1 {
+			depthCameraName := svcConfig.Sensors[1]
+			logger.Debugf("Two cameras found for slam service, assuming %v is for color and %v is for depth",
+				primarySensorName, depthCameraName)
+			depthCam, err := camera.FromDependencies(deps, depthCameraName)
+			if err != nil {
+				return "", nil, errors.Wrapf(err, "error getting camera %v for slam service", depthCameraName)
+			}
+			cams = append(cams, depthCam)
+		}
+
+		return primarySensorName, cams, nil
+	}
+	return "", nil, nil
+}
+
+// Position forwards the request for positional data to the slam library's gRPC service. Once a response is received,
+// it is unpacked into a PoseInFrame.
+func (slamSvc *builtIn) Position(ctx context.Context, name string, extra map[string]interface{}) (*referenceframe.PoseInFrame, error) {
+	ctx, span := trace.StartSpan(ctx, "slam::builtIn::Position")
+	defer span.End()
+
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return nil, err
+	}
+
+	var pInFrame *referenceframe.PoseInFrame
+	var returnedExt map[string]interface{}
+
+	// TODO: Once RSDK-1053 (https://viam.atlassian.net/browse/RSDK-1066) is complete the original code before extracting position
+	// from GetPosition will be removed and the GetPositionNew -> GetPosition
+	if slamSvc.dev {
+		slamSvc.logger.Debug("IN DEV MODE (position request)")
+		req := &pb.GetPositionNewRequest{Name: name}
+
+		resp, err := slamSvc.clientAlgo.GetPositionNew(ctx, req)
+		if err != nil {
+			return nil, errors.Wrap(err, "error getting SLAM position")
+		}
+
+		pInFrame = referenceframe.NewPoseInFrame(resp.GetComponentReference(), spatialmath.NewPoseFromProtobuf(resp.GetPose()))
+		returnedExt = resp.Extra.AsMap()
+	} else {
+		//nolint:staticcheck
+		req := &pb.GetPositionRequest{Name: name, Extra: ext}
+
+		resp, err := slamSvc.clientAlgo.GetPosition(ctx, req)
+		if err != nil {
+			return nil, errors.Wrap(err, "error getting SLAM position")
+		}
+
+		pInFrame = referenceframe.ProtobufToPoseInFrame(resp.Pose)
+		returnedExt = resp.Extra.AsMap()
+	}
+
+	// TODO DATA-531: https://viam.atlassian.net/jira/software/c/projects/DATA/boards/30?modal=detail&selectedIssue=DATA-531
+	// Remove extraction and conversion of quaternion from the extra field in the response once the Rust
+	// spatial math library is available and the desired math can be implemented on the orbSLAM side
+	if val, ok := returnedExt["quat"]; ok {
```

Comment:
[nit] should this have been refactored because we extracted the `CheckQuaternionFromClientAlgo()`?